Userland/Libraries/LibWeb/Painting/InlinePaintable.cpp

```diff
@@ -203,10 +203,50 @@ TraversalDecision InlinePaintable::hit_test(CSSPixelPoint position, HitTestType
         if (fragment_absolute_rect.contains(position_adjusted_by_scroll_offset)) {
             if (fragment.paintable().hit_test(position, type, callback) == TraversalDecision::Break)
                 return TraversalDecision::Break;
-            auto hit_test_result = HitTestResult { const_cast<Paintable&>(fragment.paintable()), fragment.text_index_at(position_adjusted_by_scroll_offset.x()) };
+            auto hit_test_result = HitTestResult { const_cast<Paintable&>(fragment.paintable()), fragment.text_index_at(position_adjusted_by_scroll_offset.x()), 0, 0 };
             if (callback(hit_test_result) == TraversalDecision::Break)
                 return TraversalDecision::Break;
         }
+
+        // If we reached this point, the position is not within the fragment. However, the fragment start or end might be the place to place the cursor.
+        // This determines whether the fragment is a good candidate for the position. The last such good fragment is chosen.
+        // The best candidate is either the end of the line above, the beginning of the line below, or the beginning or end of the current line.
+        // We arbitrarily choose to consider the end of the line above and ignore the beginning of the line below.
+        // If we knew the direction of selection, we could make a better choice.
+        auto const hit_result_outside_fragment = [fragment_absolute_rect, position_adjusted_by_scroll_offset, fragment]() -> Optional<HitTestResult> {
+            if (fragment_absolute_rect.bottom() - 1 <= position_adjusted_by_scroll_offset.y()) { // fully below the fragment
```

Comment:
Do we also need to check if we're horizontally within the fragment here?

Reply:
I believe it's covering all the cases now